library/common/engine.cc

```diff
@@ -161,22 +161,22 @@ envoy_status_t Engine::terminate() {
     }
 
     ASSERT(event_dispatcher_);
+    ASSERT(dispatcher_);
 
     // Exit the event loop and finish up in Engine::run(...)
     if (std::this_thread::get_id() == main_thread_.get_id()) {
       // TODO(goaway): figure out some way to support this.
       PANIC("Terminating the engine from its own main thread is currently unsupported.");
     } else {
       event_dispatcher_->exit();
+      dispatcher_->terminate();
```

Comment:
Is the provisional dispatcher valid after event_dispatcher_->exit() is called? Should the order here be flipped?

Reply:
I'd actually suggest moving the `exit()` call inside of `ProvisionalDispatcher::terminate()` to make this less delicate.

Reply:
Done.

Reply:
I've reverted it to what it was before. Moving the exit into `terminate()` caused the java tests to crash:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000014156e5d9, pid=28881, tid=29959
#
# JRE version: OpenJDK Runtime Environment Zulu11.50+19-CA (11.0.12+7) (build 11.0.12+7-LTS)
# Java VM: OpenJDK 64-Bit Server VM Zulu11.50+19-CA (11.0.12+7-LTS, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libndk_envoy_jni.jnilib+0x1fee5d9]  Envoy::Event::ProvisionalDispatcher::terminate()+0x29
```